[ENH]: first basic working version of the resen

asreimer opened the pull request:
This is the first working version of resen. It needs much more work, but this version allows you to:
1) create/remove a bucket
2) add/remove storage, ports, and a docker image to the bucket
3) start/stop the bucket
4) start jupyter notebook or lab server
5) list existing buckets

## Requirements
Currently, you need to have built resen-core yourself such that you have a resen-core image on your machine. For example, I built the resen-core image with the command:

    docker build -t resenlab/testing .

So the resen-core image is available with the image name "resenlab/testing".

The tool can be used in 2 ways: via a python shell, and via the resen shell. The tool requires python 3 and the docker python package.

## Testing

To test this code:

    git clone https://github.com/EarthCubeInGeo/resen.git
    cd resen
    pip install .
    cd ~
    resen

Then you will be dropped into the resen shell program:
![image](https://user-images.githubusercontent.com/4604819/57054618-91bd6900-6c4a-11e9-9356-4ff63787330b.png)

First type "help" to see the list of available commands:
![image](https://user-images.githubusercontent.com/4604819/57054713-21631780-6c4b-11e9-935b-92049bbfd2ac.png)

To see help on how to use a particular command, type "help command_name", for example, "help create_bucket":
![image](https://user-images.githubusercontent.com/4604819/57054737-49eb1180-6c4b-11e9-945b-2afdd300276d.png)

The basic workflow is:
1) create a bucket,
2) add a resencore image (for example, 'resenlab/testing' or whatever you called your resen-core build)
3) add a storage location to mount in (suggest mount an empty local directory to /home/jovyan)
4) add a port map, (suggested local 8080 bucket 8000)
5) start the bucket
6) start a jupyter notebook/lab server with ports 8080 and 8000
7) quit the jupyter server
8) stop the bucket
9) remove the bucket

Please give this a try and see if you can get through this workflow. Also, please read through the code and the comments to see the TODOs and how this tool works.

asreimer committed:
First commit of basic working resen tool.

asreimer committed:
Minor bugfixes

ljlamarche commented:
In general, this workflow seems to function well.  There are 2 main hang-up points, both involving mounting storage.
1. To add read-only storage, you need to set permissions to `ro`, not `r`.  The help entry for add_storage currently says "permissions should be 'r' or 'rw'", but setting permissions to `r` actually causes resen to crash when you issue the start_bucket command.  I think we just need to correct the Usage entry, and maybe add something in add_storage that makes sure valid permissions are set.
2. When you're mounting storage, the container_path CANNOT be /home/jovyan.  This causes whatever storage you mounted to "overwrite" the existing cashe, envs, examples, and work directories, which in turn causes problems when you try to start a jupyter notebook or a particular environment.  We should document that these are illegal file paths to mount to and maybe have the code raise an exception if you try.

asreimer committed:
Improved permissions handling. Added SElinux detection and appropriate permissions modification.

asreimer commented:
@ljlamarche 
1. This is now fixed. It accepts 'r' or 'ro' and 'rw' and automatically uses what it needs to for the underlying docker helper.
2. I've implemented a storage whitelist so that one can only into /home/jovyan/work and /home/jovyan/mount (and any location inside of those locations). This should prevent problems with mounting and accidentally wiping out files that matter.

asreimer committed:
resen can now pull the resen-core image from dockerhub

asreimer committed:
updated version number

asreimer commented:
Now the tool can pull images from dockerhub. There is a 2019.1.0rc1 version of resen-core uploaded to dockerhub here: https://hub.docker.com/r/earthcubeingeo/resen-core

Merging this in to develop now.
